### PR TITLE
pkg/metadata: Cache compiler metadata output

### PR DESCRIPTION
**Pull request by javierhonduco**
**TL;DR:** >40% of the Parca Agent CPU cycles are spent reading all the elf symbols for each binary to see whether they are stripped or not. Some extra cycles come from added pressure on the garbage collector

The compiler metadata, specifically, the stripped check, consumes a large amount of CPU as it parses and traverses the whole symbol table for each process running on the system.

This first commit caches the compiler metadata using the BuildID as the caching key.

I noticed that while checking CPU profiles in our demo environment:

image

## Test Plan

I Ran Parca and Parca Agent side by side, and everything looked good, the compiler metadata uses way fewer cycles

image

## Future work

The `string => elf.File` refactor in https://github.com/parca-dev/parca-agent/pull/876 was mostly to improve the API, performance-wise the impact of that is not very big. Will tackle this in another PR!

Signed-off-by: Francisco Javier Honduvilla Coto <javierhonduco@gmail.com>

### CODE REVIEW

File: pkg/metadata/compiler.go [hunk 1]
```diff
@@ -38,17 +53,34 @@ func Compiler() *Provider {
...
+		buildID, err := buildid.BuildID(path)
+		if err != nil {
+			return nil, fmt.Errorf("buildID failed")
+		}
+
+		value, ok := c.GetIfPresent(buildID)
+		if ok {
+			cachedLabels, ok := value.(model.LabelSet)
+			if !ok {
+				panic("The buildID cache contained the wrong type. This should never happen")
```
Comment (kakkoyun) [inline pkg/metadata/compiler.go:71]: Let's not panic. It would be better to return an error and delete the wrongly cached items.

**Review by kakkoyun — COMMENTED** [review]
LGTM. Awesome results 🎉 Just one comment, then I'm happy to merge.

File: pkg/metadata/compiler.go (hunk 1 above)
Comment (javierhonduco) [reply pkg/metadata/compiler.go:71]: For conditions that should absolutely never happen, I like panicking as it can ensure that both readers of the code understand that this is a "hard" assertion, and also ensure that if something changes in the code, later on, we have more chances of catching it. 

What do you think?

Comment (v-thakkar) [reply pkg/metadata/compiler.go:71]: I agree with Javier. I think it's good to panic here instead of returning an error.

**Comment by v-thakkar** [comment]
LGTM :tada:

**Review by brancz — APPROVED** [review]

File: pkg/metadata/compiler.go (hunk 1 above)
Comment (kakkoyun) [reply pkg/metadata/compiler.go:71]: Let's agree to disagree how we approach to panics :) It's fine for here.
